spec/models/deal_spec.rb

```diff
@@ -0,0 +1,86 @@
+require 'rails_helper'
+
+RSpec.describe Deal, type: :model do
+  subject { described_class.new(title: 'abcd', description: 'This is a random deal', price_in_cents: 10000, discount_price_in_cents: 9000, quantity: 12, tax_percentage: 10, publish_at: Time.current) }
+
+  describe 'association' do
+    it { should have_many(:deal_images).dependent(:destroy) }
+    it { should have_many(:line_items).dependent(:restrict_with_error) }
+    it { should have_many(:orders).through(:line_items) }
+    it { should accept_nested_attributes_for(:deal_images) }
+  end
+
+  describe 'validations' do
+    it 'is valid with valid attributes' do
+      expect(subject).to be_valid
+    end
+    it { should validate_presence_of(:title) }
+    it { should validate_presence_of(:description) }
+    it { should validate_presence_of(:price_in_cents) }
+    it { should validate_presence_of(:discount_price_in_cents) }
+    it { should validate_presence_of(:quantity) }
+    it { should validate_presence_of(:tax_percentage) }
+    it { should validate_presence_of(:publish_at) }
+    it { should validate_numericality_of(:discount_price_in_cents) }
+  end
+
+  describe 'callbacks' do
+    it { is_expected.to callback(:calculate_tax_on_deal).before(:validation) }
+  end
+
+  describe 'scope' do
+    describe 'all deals' do
+      context 'when no deals exist' do
+        it 'should not return any deal' do
+          expect(Deal.all_deals).to be_empty
+        end
+      end
+
+      context 'when at least one deal exist' do
+        let!(:deal) { create(:deal) }
```

Comment:
If an object is local to only one test, better to create that in that test as well inside `it-do-end` block